Jimboom7/AlphaJong
# AI-Core Rework

This is basically a whole rework of the core AI.
It seems to be running stable now (no bugs that cause the bot to crash), **BUT** it is by far not optimized yet.
Changing the old system to using the above values means that all the stuff that had been carefully finetuned before now needs to be looked at again in detail. The easiest way to do so is to let the bot play some games and watch out for moves that appear to be wrong or not optimal.
At the moment I assume the AI is equal or even slightly worse than the old AI, but after some fixes and finetuning it should become much stronger (Maybe Jade Room level).

### What is new?

- Exact shanten calculation
- Estimated hand value for own hand
- Estimated hand value for opponents hands
- More accurate estimation whether opponents are tenpai or not (Tenpai %)
- More accurate estimation whether a tile deals in or not (Deal In %)

With the above data it should be possible for the bot to make better decisions, especially in regards to Push/Fold or calling Riichi.

### How to test it

Download the new version [here](https://github.com/Jimboom7/AlphaJong/files/8682126/AlphaJong_1.3.0.user-beta.zip). It would be very helpful if you report any problems or bugs that occur!
Please use this PR for any feedback or bug reports. It would be great if you could include the log of the problematic turn (including the debug string) and/or a screenshot.

_**Currently known problems:**_
- ~~Parameters (Like EFFICIENCY_VALUE) have no effect yet, I'm still looking for a good way to integrate them.~~
- ~~Sometimes the bot calls pon or chi when there is high danger and it should already be folding.~~

**_Related Issues:_**
#36 #45 (Should improve some of the points mentioned there)
